### PR TITLE
Add signed uploads

### PR DESCRIPTION
**Pull request by bautrukevich**

### CODE REVIEW

File: src/Uploadcare/Api.php [hunk 1]
```diff
@@ -137,9 +138,18 @@ class Api
      * @param string $cdn_host CDN Host
      * @param string $cdn_protocol CDN Protocol
      * @param integer $retry_throttled Retry throttled requests this number of times
+     * @param int $lifeTime Secure signature expire time in seconds for signed uploads.
```
Comment (dmitry-mukhin) [inline src/Uploadcare/Api.php]:
```suggestion
     * @param int $lifetime Secure signature expire time in seconds for signed uploads.
```

lifetime, not life time
here and everywhere

File: src/Uploadcare/Api.php [hunk 2]
```diff
@@ -137,9 +138,18 @@ class Api
...
-    public function __construct($public_key, $secret_key, $userAgentName = null, $cdn_host = null, $cdn_protocol = null, $retry_throttled = null)
-    {
+    public function __construct(
+        $public_key,
+        $secret_key,
+        $userAgentName = null,
+        $cdn_host = null,
+        $cdn_protocol = null,
+        $retry_throttled = null,
+        $lifeTime = 0
```
Comment (dmitry-mukhin) [inline src/Uploadcare/Api.php]:
```suggestion
        $lifetime = 0
```

File: src/Uploadcare/Signature/SecureSignature.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+     * SecureSignature constructor.
+     *
+     * @param string $secretKey
+     * @param int $lifeTime
+     *
+     * @throws Exception
+     */
+    public function __construct($secretKey, $lifeTime)
+    {
+        $dateTime = new DateTime();
```
Comment (dmitry-mukhin) [inline src/Uploadcare/Signature/SecureSignature.php]:
```suggestion
        $now = new DateTime();
```

?

File: src/Uploadcare/Widget.php [hunk 4]
```diff
@@ -86,6 +105,25 @@ public function getIntegrationData()
         return $integrationData;
     }
 
+    /**
+     * Enable signed uploads.
+     *
+     * @param array $attributes Widget attributes.
+     * @return array Array with attributes.
+     */
+    private function withSignatureAttributes($attributes)
```
Comment (dmitry-mukhin) [inline src/Uploadcare/Widget.php]: why `with`?

Comment (bautrukevich) [reply src/Uploadcare/Widget.php]: Because it sounds good : ) `$this->withSignatureAttributes`. Why not?

**Comment by bautrukevich** [comment]
> 1. change varnames.

Done.

> 2. in ApiTests I don't understand what new test do except for testWidgetSignedUploads

Just want to make sure that signed uploads work correctly with all upload methods

File: src/Uploadcare/Widget.php (hunk 4 above)
Comment (dmitry-mukhin) [reply src/Uploadcare/Widget.php]: `addSignagureAttributes` makes more sense here

**Comment by dmitry-mukhin** [comment]
in ApiTests you're catching exceptions, and check that exceptions has some proper attrs.
AFAIU, this isn't checking that uploads work.

**Comment by bautrukevich** [comment]
@dmitry-mukhin Yes, you are right. But maybe you have some ideas on how to test this and not to make breaking changes into the API?

**Comment by bautrukevich** [comment]
As a variant, we can test with some credentials of the project that have signed uploads enabled. But I think that it's enough to check only that we pass signature and expire (because we test them already)

File: src/Uploadcare/Widget.php (hunk 4 above)
Comment (dmitry-mukhin) [reply src/Uploadcare/Widget.php]: ^

**Pull request merged**